File: Core-DSA-Optimized/Strings/KMP-Search/kmp.py
class KMPSearch:
    @staticmethod
    def _compute_lps(pattern: str) -> list[int]:
        m = len(pattern)
        lps = [0] * m
        length = 0
        i = 1

        while i < m:
            if pattern[i] == pattern[length]:
                length += 1
                lps[i] = length
                i += 1
            else:
                if length != 0:
                    length = lps[length - 1]
                else:
                    lps[i] = 0
                    i += 1
        return lps

    @staticmethod
    def search(text: str, pattern: str) -> list[int]:
        n = len(text)
        m = len(pattern)
        if m == 0: return []
        
        lps = KMPSearch._compute_lps(pattern)
        indices = []
        i = 0 # index for text
        j = 0 # index for pattern

        while i < n:
            if pattern[j] == text[i]:
                i += 1
                j += 1

            if j == m:
                indices.append(i - j)
                j = lps[j - 1]
            elif i < n and pattern[j] != text[i]:
                if j != 0:
                    j = lps[j - 1]
                else:
                    i += 1
                    
        return indices

# --- CI/CD Automated Test ---
if __name__ == '__main__':
    dna_sequence = "GACACCCTACACAACCACCACACACCCCCACACAC"
    virus_signature = "ACAC"
    
    matches = KMPSearch.search(dna_sequence, virus_signature)
    
    assert len(matches) == 5, f"Expected 5 matches, found {len(matches)}"
    assert matches == [1, 9, 20, 29, 31], "Incorrect match indices found!"
    
    print("Python KMP Search Test Passed!")

File: Core-DSA-Optimized/Strings/KMP-Search/kmp.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cassert>

using namespace std;

class KMPSearch {
private:
    static vector<int> computeLPS(const string& pattern) {
        int m = pattern.length();
        vector<int> lps(m, 0);
        int len = 0;
        int i = 1;

        while (i < m) {
            if (pattern[i] == pattern[len]) {
                len++;
                lps[i] = len;
                i++;
            } else {
                if (len != 0) {
                    len = lps[len - 1];
                } else {
                    lps[i] = 0;
                    i++;
                }
            }
        }
        return lps;
    }

public:
    static vector<int> search(const string& text, const string& pattern) {
        vector<int> indices;
        int n = text.length();
        int m = pattern.length();
        if (m == 0) return indices;

        vector<int> lps = computeLPS(pattern);
        int i = 0;
        int j = 0;

        while (i < n) {
            if (pattern[j] == text[i]) {
                i++;
                j++;
            }
            if (j == m) {
                indices.push_back(i - j);
                j = lps[j - 1];
            } else if (i < n && pattern[j] != text[i]) {
                if (j != 0) {
                    j = lps[j - 1];
                } else {
                    i++;
                }
            }
        }
        return indices;
    }
};

// --- CI/CD Automated Test ---
int main() {
    string dnaSequence = "GACACCCTACACAACCACCACACACCCCCACACAC";
    string virusSignature = "ACAC";
    
    vector<int> matches = KMPSearch::search(dnaSequence, virusSignature);
    
    assert(matches.size() == 5);
    assert(matches[0] == 1);
    assert(matches[4] == 31);
    
    cout << "C++ KMP Search Test Passed!\n";
    return 0;
}